NumberFormattingUtils

danimoh opened the pull request.

nibhar commented on src/number-formatting-utils/NumberFormattingUtils.ts:
```diff
@@ -0,0 +1,139 @@
+function _disassembleNumber(
+    value: string | number | bigint | BigInteger,
+): { sign: '-' | '', integerPart: string, dotPosition: number, decimalPart: string } {
+    value = toNonScientificNumberString(value);
```
`toNonScientificNumberString` also does the regular expression matching. I think the regular expression is matched at the very least twice for input strings, potentially (formatNumber with scientific notation and maxDigits set) even 6 times. I feel like there is room for improvement here (maybe an intermediate class representation or something like that).

danimoh replied:
I also thought about that and considered passing the disassembled objects between classes.
But at the end, keeping it simple is nicer in my opinion as string parsing should be very fast and these methods are not called too often.

nibhar replied:
I personally think that a i.e `FormattableNumber` class which works on the disassembled Number is not too complicated. 
As all the functions work on the disassembled object anyways I think an intermediary representation makes more sense. A format number could then look like this:

```typescript
class FormattableNumber { ... }
export function formatNumber(
    value: string | number | bigint | BigInteger,
    maxDecimals?: number,
    minDecimals?: number,
    groupSeparator: string = '\u202F', // thin space (U+202F)
): string {
  const formattableNumber = new FormattableNumber(value); // disassembles
  formattableNumber.normalize(); // removes scientific notation
  return formattableNumber(maxDecimals, minDecimals, groupSeparator);
}
```

danimoh replied:
I switched to your suggested class based approach. It is in fact nicer :+1:

nibhar commented on src/number-formatting-utils/NumberFormattingUtils.ts:
```diff
@@ -0,0 +1,139 @@
+}
+
+// moves comma and returns a string without precision loss
+export function moveComma(
```
This does not have the option to move the comma and restrict the result to having a set amount of decimals at the same time, similar to what `formatNumber` does. As I think that would be a strong use case for this function I think that it might make sense to add that functionality, in order to not have to use `formatNumber(moveComma(value, -9), 5, 2);`

danimoh replied:
I don't think that mixing these functionalities up is a good idea.

danimoh commented on src/number-formatting-utils/NumberFormattingUtils.ts:
```diff
@@ -0,0 +1,139 @@
+}
+
+// moves comma and returns a string without precision loss
+export function moveComma(
```
You have a suggestion?

nibhar replied:
could be i.e `timesPowerOfTen(n: value)`, `moveDecimalPosition(n: number)`.

It might be better to get rid of it entirely and work on the caller side with arithmetic, such that the input to `formatNumber` is already correct.

danimoh replied:
Renamed to `moveDecimalSeparator` now.
Note that this method is in fact required to resolve numbers with scientific notation and can not be simply replaced by arithmetics on the caller side which can result in scientific notaion on `toString`. Also this method is free of precision issues and can work on strings.
Also the arithmetics on `BigInteger` instances already result in as much code as this helper method and would be required everywhere we want to transform a `BigInteger`.

nibhar replied:
Yes, I ran into that problem myself already so I see the point of having it.

danimoh commented:
I changed the implementation to your class based approach with some changes, please have a look.
I also switched to exporting the whole class instead of single helper methods.
I exported single helper methods before for better tree-shakeability.
But now with the class based approach I prefer exporting the whole class as it is likely not well tree-shaekable anymore anyways.
I will adapt the use in `vue-components` and the `hub` tomorrow.

I also added an `equals` method which can be useful for your requested approximation check.

nibhar commented on src/formattable-number/FormattableNumber.ts:
```diff
@@ -0,0 +1,110 @@
+export function toNonScientificNumberString(
+    value: string | number | bigint | BigInteger,
+): string {
+    return new FormattableNumber(value).render({ groupSeparator: false });
```
```suggestion
    return new FormattableNumber(value).render();
```

danimoh replied:
This is not the same. The default for `groupSeparator` is `'\u202F'`.
However, I noticed when adapting code occurences to the new API, that usages with `groupSeparator: false` outweigh `groupSeparator: '\u202F'` by far.
So maybe I should switch the default to `false` (or `undefined` in which case also the `false` can be removed from the api).
Would require changing all calls to `render` again.

sisou replied:
In this case, with the one call happening more often, vs the other with a hard to remember unicode code, I think it's better to separate them simply into two methods. One that renders without and one that renders with separator. For example `render()` and `renderSeparated()` or `renderGrouped()`. 

Maybe `toString()` is a better method name even? (together with `toGroupedString()`.

danimoh replied:
I like `toString` :+1:
However, I don't like boating the API with additional methods that basically do the same. Instead I added a `useGrouping` property to the option object.
If there are no complains, I will go ahead and merge this PR into master.

danimoh commented on src/formattable-number/FormattableNumber.ts:
```diff
@@ -0,0 +1,110 @@
+    }: {
+        maxDecimals?: number,
+        minDecimals?: number,
+        groupSeparator?: string | false,
```
It's not necessary. When being used as in line 109, `groupSeparator: false` just reads nicer than `groupSeparator: ''`.

danimoh replied:
The newest changes now made `false` obsolete as option and I removed it.

sisou replied:
With rebasing, changes are not traceable. What did you change how?

danimoh replied:
Github shows what changed on a force push: https://github.com/nimiq/nimiq-utils/compare/adcadaad4f654ba556c81f4d301c37d8f8fb7b5c..00fd0893f376f4573deb3f19bc6bf9a0701e1674

sisou commented on src/formattable-number/FormattableNumber.ts:
```diff
@@ -0,0 +1,116 @@
+        if (exponent) this.moveDecimalSeparator(exponent); // remove scientific notation
+    }
+
+    public toString({
```
One last possibility is to accept as an argument either an object _or a boolean_. The boolean representing the `useGrouping` option. Meaning that instead of `toString({useGrouping: true})` I could just do `toString(true)` if I don't want to change anything else.

danimoh replied:
This is now supported for if somebody wants to use it :D

sisou commented:
On mobile it doesn't, and on desktop it only shows for the last recent force push, as far as I know.


Individual commits while working on it are better, and then consolidate them all together when actually merging.


I thought you preferred that too?


via Newton Mail [https://cloudmagic.com/k/d/mailapp?ct=pa&cv=10.0.25&pv=9&source=email_footer_2] On Sat, Oct 26, 2019 at 6:02pm, danimoh < notifications@github.com [notifications@github.com] > wrote:
@danimoh commented on this pull request.




In src/formattable-number/FormattableNumber.ts [https://github.com/nimiq/nimiq-utils/pull/28#discussion_r339305363] :

> +        this._digits = `${numberMatch[2]}${numberMatch[3]}`;
+        if (!this._digits) throw new Error(`${value} is not a valid number`);
+        this._decimalSeparatorPosition = numberMatch[2].length;
+
+        const exponent = Number.parseInt(numberMatch[5], 10);
+        if (exponent) this.moveDecimalSeparator(exponent); // remove scientific notation
+    }
+
+    public render({
+        maxDecimals = undefined,
+        minDecimals = undefined,
+        groupSeparator = '\u202F', // thin space (U+202F)
+    }: {
+        maxDecimals?: number,
+        minDecimals?: number,
+        groupSeparator?: string | false,


Github shows what changed on a force push: https://github.com/nimiq/nimiq-utils/compare/adcadaad4f654ba556c81f4d301c37d8f8fb7b5c..00fd0893f376f4573deb3f19bc6bf9a0701e1674 [https://github.com/nimiq/nimiq-utils/compare/adcadaad4f654ba556c81f4d301c37d8f8fb7b5c..00fd0893f376f4573deb3f19bc6bf9a0701e1674]

—
You are receiving this because your review was requested.
Reply to this email directly, view it on GitHub [https://github.com/nimiq/nimiq-utils/pull/28?email_source=notifications&email_token=AAN6KQ2H3J54V2WGKBO4GUDQQRS25A5CNFSM4IYAV4EKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCJKJCPA#discussion_r339305363] , or unsubscribe [https://github.com/notifications/unsubscribe-auth/AAN6KQ73PETHNPLCPSNQKEDQQRS25ANCNFSM4IYAV4EA] .

danimoh commented:
> On mobile it doesn't, and on desktop it only shows for the last recent force push, as far as I know.

I'm not sure about its behavior on repeated force pushs. It might also diff between the state before the first force push and the state after the last one?

> Individual commits while working on it are better, and then consolidate them all together when actually merging.
> I thought you preferred that too?

When a requested change only results in a small change in a previous commit, I often prefer to rebase.
